src/transformStateWithClones.js

```diff
@@ -7,7 +7,46 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const STATE_CLONE = Object.assign({}, state);
+  const RESULT_ARR = [];
+
+  for (const action of actions) {
+    const ACTION_TYPE = action.type;
+
+    switch (ACTION_TYPE) {
+      case 'addProperties':
+        addProperties(STATE_CLONE, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(STATE_CLONE, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clearProperties(STATE_CLONE);
+        break;
+    }
+
+    RESULT_ARR.push({ ...STATE_CLONE });
+  }
+
+  return RESULT_ARR;
+
+  function addProperties(changingObject, extraData) {
```

Comment:
generally, you really use it only once, but if you move the logic to the function, you better declare the function outside the `transformStateWithClones` function